### PR TITLE
Allows user to update their avatar

### PR DESCRIPTION
**Pull request by andrewxhill**

### CODE REVIEW

File: App/SB/components/ThreadDetailCard/index.js [hunk 1]
```diff
@@ -60,7 +60,11 @@ const ThreadDetailCard = props => {
       }
 
       const defaultSource = require('../../views/Settings/statics/main-image.png')
-      const uri = props.photo.author_id ? 'https://cafe.us-east-1.textile.io/ipns/' + props.photo.author_id + '/avatar' : undefined
+      let uri = props.photo.author_id ? 'https://cafe.us-east-1.textile.io/ipns/' + props.photo.author_id + '/avatar' : undefined
+      // ensure we have the user's latest avatar even if the cafe is still caching
+      if (props.profile && props.profile.id && props.profile.id === props.photo.author_id) {
+        uri = 'https://cafe.us-east-1.textile.io' + props.profile.avatar_id
```
Comment (asutula) [inline App/SB/components/ThreadDetailCard/index.js]: I thought `props.profile.avatar_id` is actually a full URL. It's confusing, we should update that name.

File: App/SB/components/ThreadListCard/index.js [hunk 2]
```diff
@@ -13,6 +13,12 @@ const ThreadCard = props => {
   const lastUpdatedTime = moment(updated).fromNow()
   const commentsCount = photos.length
   let uri = latestPeerId ? 'https://cafe.us-east-1.textile.io/ipns/' + latestPeerId + '/avatar' : undefined
+
+  // ensure we have the user's latest avatar even if the cafe is still caching
+  if (props.profile && props.profile.id && props.profile.id === latestPeerId) {
+    uri = 'https://cafe.us-east-1.textile.io' + props.profile.avatar_id
```
Comment (asutula) [inline App/SB/components/ThreadListCard/index.js]: Same as above.

File: App/SB/views/UserProfile/index.js [hunk 3]
```diff
@@ -38,14 +36,14 @@ class UserProfile extends React.PureComponent {
...
         <HeaderButtons>
-          <Item 
+          <Item
             title='Avatar'
             buttonWrapperStyle={{marginLeft: 11, marginRight: 11}}
             ButtonElement={
               <Avatar
-                width={32} 
-                height={32} 
-                uri={params.avatarUrl}
```
Comment (asutula) [inline App/SB/views/UserProfile/index.js]: Just set my editor to trim trailing whitespace, should get rid of all this PR noise with us.

File: App/SB/views/UserProfile/index.js [hunk 4]
```diff
@@ -156,7 +164,8 @@ const mapStateToProps = (state, ownProps) => {
 
 const mapDispatchToProps = (dispatch) => {
   return {
-    lockScreen: () => { dispatch(AuthActions.logOutRequest()) }
+    lockScreen: () => { dispatch(AuthActions.logOutRequest()) },
+    changeAvatar: () => { dispatch(PreferencesActions.changeAvatarRequest()) }
```
Comment (asutula) [inline App/SB/views/UserProfile/index.js]: Don't see where this is called. Do we need `PreferencesActions.changeAvatarRequest` to exist?

File: App/Sagas/TextileSagas.ts [hunk 5]
```diff
@@ -138,6 +138,53 @@ export function * handleProfilePhotoSelected(action: ActionType<typeof UIActions
   }
 }
 
+export function * handleProfilePhotoUpdated(action: ActionType<typeof UIActions.updateProfilePicture>) {
```
Comment (asutula) [inline App/Sagas/TextileSagas.ts]: Looks like lots of shared code with the `handleProfilePhotoSelected` method. Should we combine?

File: App/SB/components/ThreadDetailCard/index.js (hunk 1 above)
Comment (andrewxhill) [reply App/SB/components/ThreadDetailCard/index.js]: just reflects the key in go, @sanderpick is there a tie to avatar_id?

File: App/SB/views/UserProfile/index.js (hunk 3 above)
Comment (andrewxhill) [reply App/SB/views/UserProfile/index.js]: excellent

File: App/SB/components/ThreadDetailCard/index.js (hunk 1 above)
Comment (asutula) [reply App/SB/components/ThreadDetailCard/index.js]: I'm just saying that I think your concat of `https://cafe.us-east-1.textile.io` with `props.profile.avatar_id` will result in ``https://cafe.us-east-1.textile.iohttps://cafe.us-east-1.textile.io/someId`

Comment (andrewxhill) [reply App/SB/components/ThreadDetailCard/index.js]: avatar_id looks something like, `/ipfs/QmWUnMBtSRzaMTj622aHESaRtHLdbu93XvxSMug31VvGkT/thumb?key=18lJGIOSnCS7xA3uBAFEuAEdEHD18lJGOWJfbsZl5PuA`

Comment (asutula) [reply App/SB/components/ThreadDetailCard/index.js]: Ah ok, that new. Didn't know.

File: App/Redux/UIRedux.ts [hunk 6]
```diff
@@ -11,6 +11,15 @@ const actions = {
   selectProfilePicture: createAction('SELECT_PROFILE_PICTURE', resolve => {
     return (uri: string) => resolve({ uri })
   }),
+  updateProfilePicture: createAction('UPDATE_PROFILE_PICTURE', resolve => {
+    return (uri: string) => resolve({ uri })
+  }),
+  updateProfileSuccess: createAction('UPDATE_PROFILE_PICTURE_SUCCESS', resolve => {
```
[the remark this answers is not in the document]
Comment (andrewxhill) [reply App/Redux/UIRedux.ts]: it is as of 10s ago, https://github.com/textileio/textile-mobile/pull/355/files#diff-f5f46a6afa92bf9d61f6f162163d925bR107

[the remark this answers is not in the document]
Comment (andrewxhill) [reply App/Redux/UIRedux.ts]: no, nm you were right to remove it. gone

File: App/Sagas/TextileSagas.ts [hunk 7]
```diff
@@ -89,24 +89,38 @@ export function * handleProfilePhotoSelected(action: ActionType<typeof UIActions
...
+  yield processAvatarImage(action.payload.uri, defaultThread)
+}
+
+export function * handleProfilePhotoUpdated(action: ActionType<typeof UIActions.updateProfilePicture>) {
+  yield call(NavigationService.navigate, 'TabNavigator')
+
+  let defaultThread: TextileTypes.Thread = yield call(getDefaultThread)
 
-  const photoPath = action.payload.uri.replace('file://', '')
+  yield processAvatarImage(action.payload.uri, defaultThread)
```
Comment (asutula) [inline App/Sagas/TextileSagas.ts]: Ok probably last thing, this should be `yield * AvatarImage...`. I think you can also use `yield call()` to run other generator functions too. Not sure what the diff is.

File: App/Sagas/TextileSagas.ts [hunk 8]
```diff
@@ -89,24 +89,38 @@ export function * handleProfilePhotoSelected(action: ActionType<typeof UIActions
...
-      defaultThread = action.payload.thread
-      yield put(ThreadsActions.refreshThreadsRequest())
-    }
+  if (!defaultThread) {
+    yield put(ThreadsActions.addThreadRequest('default'))
+    const action: ActionType<typeof ThreadsActions.addThreadSuccess> = yield take(getType(ThreadsActions.addThreadSuccess))
+    defaultThread = action.payload.thread
+    yield put(ThreadsActions.refreshThreadsRequest())
+  }
+  yield processAvatarImage(action.payload.uri, defaultThread)
```
Comment (asutula) [inline App/Sagas/TextileSagas.ts]: Same fix here.

Comment (andrewxhill) [reply App/Sagas/TextileSagas.ts]: didn't realize, done

**Comment by asutula** [comment]
Nice.

**Pull request merged**